src/variable.rs

```diff
@@ -66,580 +111,594 @@ pub trait Numeric {
+                    let _g = LOCK.lock().unwrap();
+                    err = $nc_get_var1_type(variable.ncid, variable.varid, indices_ptr, &mut buff);
                 }
                 if err != NC_NOERR {
-                    return Err(NC_ERRORS.get(&err).unwrap().clone());
+                    return Err(err.into());
                 }
                 Ok(buff)
             }
-            
-            // fetch a SLICE of values from variable using `$nc_get_vara`
-            fn slice_from_variable(variable: &Variable, indices: &[usize], slice_len: &[usize]) -> Result<Vec<$sized_type>, String> {
-                // Check the length of `indices`
-                if indices.len() != variable.dimensions.len() {
-                    return Err("`indices` must has the same length as the variable dimensions".into());
-                }
-                if indices.len() != slice_len.len() {
-                    return Err("`slice` must has the same length as the variable dimensions".into());
-                }
-                let mut values: Vec<$sized_type>;
-                let mut values_len: usize = 1;
-                for i in 0..indices.len() {
-                    if (indices[i] as u64) >= variable.dimensions[i].len {
-                        return Err("requested index is bigger than the dimension length".into());
-                    }
-                    if ((indices[i] + slice_len[i]) as u64) > variable.dimensions[i].len {
-                        return Err("requested slice is bigger than the dimension length".into());
-                    }
-                    // Compute the full size of the request values
-                    if slice_len[i] > 0 {
-                        values_len *= slice_len[i];
-                    } else {
-                        return Err("Each slice element must be superior than 0".into());
+
+            #[cfg(feature = "ndarray")]
+            fn array_from_variable(
+                variable: &Variable,
+                indices: Option<&[usize]>,
+                slice_len: Option<&[usize]>,
+            ) -> error::Result<ArrayD<$sized_type>> {
+                let _slice_len: Vec<_>;
+                let slice_len = match slice_len {
+                    Some(x) => x,
+                    None => {
+                        _slice_len = variable.dimensions.iter().map(|x| x.len()).collect();
+                        &_slice_len
                     }
-                }
+                };
 
-                let err: i32;
-                // Get a pointer to an array [size_t]
-                let indices: Vec<size_t> = indices.iter().map(|i| *i as size_t).collect();
-                let slice: Vec<size_t> = slice_len.iter().map(|i| *i as size_t).collect();
-                unsafe {
-                    let _g = libnetcdf_lock.lock().unwrap();
+                let mut values: ArrayD<$sized_type> = unsafe { ArrayD::uninitialized(slice_len) };
```

Comment:
I can't recommend `uninitialized` anymore unfortunately. It will be deprecated, but it hasn't happened yet.

This is due to the developing consensus about unsafe coding guidelines and how to handle uninitialized values (this development is not yet finished).

What happens here that will be discouraged is that a `&mut [T]` is taken where the contents is uninitialized, while a raw pointer would be the only good way to do that (so a `*mut [T]` or `*mut T` and length).

I don't have a great solution, but I imagine `Vec::with_capacity` to allocate an uninit buffer and `v.as_mut_ptr()` to get a raw pointer would be the way.

Later, when the vec data is initialized we can use `Array::from(v).into_shape(slice_len).unwrap()`.

Reply:
I went with `Vec::with_capacity()` and `Vec::set_len`. Are the official docs for `Vec::set_len` up to date regarding the latest unitialized progress?

Reply:
Yes! Love it.

You can see a fun trace of the ongoing work - [`Vec::as_mut_ptr`](https://doc.rust-lang.org/nightly/std/vec/struct.Vec.html#method.as_mut_ptr) is now its own method and marked stable since *Rust 1.37* in the docs. (The old version is `.as_mut_ptr` through deref to slice).

This signifies that we can fetch the raw pointer from the Vec without going through a slice.

Reply:
That is a very interesting piece!